# SOLUTIONS AND EXPLANATIONS FOR LESSON 1: MATRIX FOR AI (20/6/2026) HOMEWORK - USING TEXTS

## PREPARATION FOR BOTH EXERCISES: TRANSFORM TEXTS TO VECTORS AND MATRIXES

Import numpy lib and create a list of 8 sentences to transform to vectors

In [99]:
import numpy as np
sentence = ["i love you", 
            "this house is big",
            "you are beautiful",
            "that will do",
            "you are welcome", 
            "how are you", 
            "what is your name", 
            "i play badminton"]

extract each separate word from every sentence and then sort them in alphabet order

In [100]:
word = []
for s in sentence:
    for w in s.lower().split():
        word.append(w)
word.sort()
print(word)

['are', 'are', 'are', 'badminton', 'beautiful', 'big', 'do', 'house', 'how', 'i', 'i', 'is', 'is', 'love', 'name', 'play', 'that', 'this', 'welcome', 'what', 'will', 'you', 'you', 'you', 'you', 'your']


building a function to transform these words to vectors

In [101]:
def to_vectors(s):
    v = np.zeros(len(word)) # an array all 0 with length of word list above
    for w in s.lower().split():
        v[word.index(w)] += 1 
    return v
matrix = np.array([to_vectors(s) for s in sentence], dtype=int)
print(matrix)
print(matrix.shape)

[[0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0]
 [0 0 0 0 0 1 0 1 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 0 0 0]
 [1 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0]
 [0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0 0]
 [1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0]
 [1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0 1 0 0 0 0 0 1]
 [0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0]]
(8, 26)


Here, 8 sentences are separately transformed to 8 vectors (each with 26 dims or 26 words in total) which makes a 8x26 matrix.
- 8 rows are 8 sentences whose "1" stands at the index corresponds to a word in the sorted word list above
- Each column represents the frequency that each word appears in each sentence. for ex, the first column represents for word[0] ('are') that 'are' appears in sentence 3,5 and 6 one time.

## EXERCISE 1: CALCULATE THE SIMILARITY

### Basic operation: Mean-centering

We take the mean of each column

In [102]:
col_mean = np.mean(matrix, axis=0, keepdims=True) # axis=0 means that we are taking average by columns (=1 is for rows)
print(col_mean)
print(col_mean.shape)

[[0.375 0.    0.    0.125 0.125 0.125 0.125 0.125 0.125 0.25  0.    0.25
  0.    0.125 0.125 0.125 0.125 0.125 0.125 0.125 0.125 0.5   0.    0.
  0.    0.125]]
(1, 26)


first column average = 3/8 = 0.375. Here we keep the original dims with 26 columns

Now, mean centering

In [103]:
centered_matrix = matrix - col_mean
print(centered_matrix)
print(centered_matrix.shape)

[[-0.375  0.     0.    -0.125 -0.125 -0.125 -0.125 -0.125 -0.125  0.75
   0.    -0.25   0.     0.875 -0.125 -0.125 -0.125 -0.125 -0.125 -0.125
  -0.125  0.5    0.     0.     0.    -0.125]
 [-0.375  0.     0.    -0.125 -0.125  0.875 -0.125  0.875 -0.125 -0.25
   0.     0.75   0.    -0.125 -0.125 -0.125 -0.125  0.875 -0.125 -0.125
  -0.125 -0.5    0.     0.     0.    -0.125]
 [ 0.625  0.     0.    -0.125  0.875 -0.125 -0.125 -0.125 -0.125 -0.25
   0.    -0.25   0.    -0.125 -0.125 -0.125 -0.125 -0.125 -0.125 -0.125
  -0.125  0.5    0.     0.     0.    -0.125]
 [-0.375  0.     0.    -0.125 -0.125 -0.125  0.875 -0.125 -0.125 -0.25
   0.    -0.25   0.    -0.125 -0.125 -0.125  0.875 -0.125 -0.125 -0.125
   0.875 -0.5    0.     0.     0.    -0.125]
 [ 0.625  0.     0.    -0.125 -0.125 -0.125 -0.125 -0.125 -0.125 -0.25
   0.    -0.25   0.    -0.125 -0.125 -0.125 -0.125 -0.125  0.875 -0.125
  -0.125  0.5    0.     0.     0.    -0.125]
 [ 0.625  0.     0.    -0.125 -0.125 -0.125 -0.125 -0.125  0

Numpy will automatically copy the col_mean matrix above to 7 more rows to make a 8x26 matrix (same shape as matrix). This is called broadcasting in Numpy.  
Subtracting each entry with the average will make each column's mean = 0.

In [104]:
column_mean = np.mean(centered_matrix, axis=0, keepdims=True)
print(column_mean)

[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0.]]


### Cosine similarity

is basically comparing the value of cosine between 2 vectors:
- ~1 means they are highly related

In [105]:
def cosine_similarity(X, Y = None):
    if Y is None:
        Y = X
    
    Xn = X / np.linalg.norm(X, axis=1, keepdims=True) # divided by the length (norm) of itself or Vector Normalization
    Yn = Y / np.linalg.norm(Y, axis=1, keepdims=True)

    return Xn @ Yn.T

This function corresponds to the formula: $cos = \frac{u \times v}{||u|| \times ||v||}$

In [106]:
# Calculate with sentence 1 and 2
print(f'cosine similarity of sentence 1 and 2 is: {cosine_similarity(matrix[0:1], matrix[1:2])}')

# Calculate with sentence 1 and 8
print(f'cosine similarity of sentence 1 and 8 is: {cosine_similarity(matrix[0:1], matrix[7:8])}')

# Calculate with sentence 3 and 5
print(f'cosine similarity of sentence 3 and 5 is: {cosine_similarity(matrix[2:3], matrix[4:5])}')


cosine similarity of sentence 1 and 2 is: [[0.]]
cosine similarity of sentence 1 and 8 is: [[0.33333333]]
cosine similarity of sentence 3 and 5 is: [[0.66666667]]


sentence 1 and 2 have 0/3 identical words. Cos = 0 means 2 vectors are perpendicular and are not related  
whereas 1 and 8 have 1/3 and 3 and 5 have 2/3

### Word Embedding - Semantic research

The to_vector method above is just a way to vectorize a sentence to count the appearance of words. It does not account for the semantic or meaning of them.  
For example, 'I go by bike' and 'I ride a bicycle' would end up having a very ~0 cosine value as the function doesn't know that bike and bicycle are synonyms or go by and ride is almost the same.


This is where Semantic search is introduced.  
First, engineers create a vector that has 100-768 dims (most LLMs now choose 768) representing for a separate word. Each dim is a randomly generated number which show a property of that word.  
Then the model will go through 2 games: masked language modelling and next sentence prediction to adjust those numbers so that each word will have its own set and the relations between vocabs are precise. For ex, 'cat' and 'dog' will be closed not 'cat' and 'car'.  
Finally, when users input, the model will use cosine similarity to find the top-related words in terms of context and meaning.

The same goes for sentences. Here we will simulate the algorithm by generating a random 1x768 matrix for each sentence.

In [107]:
train = np.random.rand(len(sentence), 768)
print(train)
print(train.shape)

[[0.75832875 0.82485291 0.76462226 ... 0.85571345 0.59595864 0.57247582]
 [0.78093413 0.07839706 0.54461552 ... 0.22137904 0.27786295 0.42131532]
 [0.90206702 0.06451482 0.92265495 ... 0.47935919 0.99507689 0.77877425]
 ...
 [0.50604874 0.79937532 0.98068274 ... 0.5226429  0.57816061 0.71451734]
 [0.96931485 0.15796566 0.47382238 ... 0.33552526 0.40185065 0.86564684]
 [0.82952582 0.52595844 0.42093079 ... 0.86192519 0.23260307 0.77991519]]
(8, 768)


create a semantic search function

In [108]:
def search(query, top_k=3):
    q = np.random.rand(1, 768) # transform the user input also to a 1x768

    cos = cosine_similarity(q, train) # thanks to broadcasting, numpy will calculate the cosine similarity of q to each of our sentence
    cos = cos.flatten() # change to 1D (array) for easier cals
    print(cos) # will be random each run as we choose random 768 numbers

    #In order to sort from small to big cos value without losing track of which value correspond to which sentence -> use argsort
    cos_idx_sort = np.argsort(cos)
    print(cos_idx_sort)

    T_k = cos_idx_sort[-top_k:][::-1] # Take out top 3 and then sort backwards to get big -> small
    print(T_k)

    return [(sentence[x], round(float(cos[x]), 3)) for x in T_k]

search('love you')

[0.75795629 0.74436239 0.75884608 0.74712854 0.76098131 0.75876993
 0.76525098 0.74808222]
[1 3 7 0 5 2 4 6]
[6 4 2]


[('what is your name', 0.765),
 ('you are welcome', 0.761),
 ('you are beautiful', 0.759)]

In [109]:
search('skibidi toilet')

[0.75724546 0.75383248 0.74627242 0.74631292 0.75152461 0.73976986
 0.75801818 0.75015492]
[5 2 3 7 4 1 0 6]
[6 0 1]


[('what is your name', 0.758),
 ('i love you', 0.757),
 ('this house is big', 0.754)]

The most related pairs will have the cosine value ~1.  
The results are not close and correct as we think, cuz all the 768 numbers are random and our word-embedding is nowhere near the reality.

## EXERCISE 2: LINEAR TRANSFORMATION & SVD

SVD is a way to decompose a matrix into 3 components that can show the variance of data (referencing our major assignment in Linear Algebra - Randomized PCA).  
centered_matrix = $U \Sigma V^\top$

In [111]:
U, Sigma, Vt = np.linalg.svd(centered_matrix)
print(U.shape)
print(Sigma.shape)
print(Vt.shape)

(8, 8)
(8,)
(26, 26)


Then we truncate each components down to 2 dims

In [ ]:
U_k = U[:,:2]
print(U_k)
print(U_k.shape)

Sigma_k = Sigma[:2]
print(Sigma_k)
print(Sigma_k.shape)

[[ 0.16130518  0.32732684]
 [-0.50148896 -0.32732684]
 [ 0.37743107 -0.21821789]
 [-0.17181994  0.32732684]
 [ 0.37743107 -0.21821789]
 [ 0.37743107 -0.21821789]
 [-0.50148896 -0.32732684]
 [-0.11880051  0.65465367]]
(8, 2)
[2.45812499 2.        ]
(2,)


Here, we don't use Vt because if multiply, it will become the original with separate words.  
We just want to embed 8 sentences + 8 topics and their weights.

In [121]:
Sigma_k = np.diag(Sigma_k) # transform Sigma_k from 1D array to square matrix
sentence_embed = U_k @ Sigma_k
print(sentence_embed)

[[ 0.39650828  0.65465367]
 [-1.23272255 -0.65465367]
 [ 0.92777274 -0.43643578]
 [-0.42235489  0.65465367]
 [ 0.92777274 -0.43643578]
 [ 0.92777274 -0.43643578]
 [-1.23272255 -0.65465367]
 [-0.29202651  1.30930734]]


Now, each sentence has been cut down to 2 dims presenting their 2 most significant features/topics.  
We will sketch this to a graph to demonstrate.